[scroll-animations] Implement view-timeline-inset property

nmoucht opened the pull request:
#### 653c136096bd0586685b8ec24ff51682e1b6c3d9
<pre>
[scroll-animations] Implement view-timeline-inset property
<a href="https://bugs.webkit.org/show_bug.cgi?id=280655">https://bugs.webkit.org/show_bug.cgi?id=280655</a>
<a href="https://rdar.apple.com/137009069">rdar://137009069</a>

Reviewed by Antoine Quint.

Implement parsing of view timeline inset if set via the ViewTimelineOptions,
and adjust computeViewTimelineData to use insets.

* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::lengthForInset):
(WebCore::insetsFromString):
(WebCore::insetsFromOptions):
(WebCore::ViewTimeline::ViewTimeline):
(WebCore::ViewTimeline::computeViewTimelineData const):
(WebCore::ViewTimeline::currentTime):
* Source/WebCore/animation/ViewTimeline.h:

Canonical link: <a href="https://commits.webkit.org/284636@main">https://commits.webkit.org/284636@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e984231061b89d325ebba3abc5438da9e9b2eb2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70040 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49443 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22798 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74127 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21206 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72157 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57245 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21049 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/55588 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14061 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73106 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45044 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60415 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36054 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41704 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17846 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19575 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63628 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18193 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75844 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14265 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17427 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63287 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14305 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60484 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63215 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11230 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4850 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10696 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45248 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46322 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47596 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46063 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -74,13 +78,56 @@ Ref<ViewTimeline> ViewTimeline::createFromCSSValue(Style::BuilderState& builderS
+    if (value->isPercentage())
+        return Length(value->resolveAsPercentageNoConversionDataRequired(), LengthType::Percent);
+    if (value->isPx())
+        return Length(value->resolveAsLengthNoConversionDataRequired<int>(), LengthType::Fixed);
```
Evidently this leads to a crash in `imported/w3c/web-platform-tests/scroll-animations/view-timelines/unattached-subject-inset.html` due to an `!isCalculated()` assertion. Not sure why we're hitting this since that test doesn't seem to use `calc()`.

nmoucht replied:
This test does actually use CSSMathNegate which might be causing this. I'm confused as to how we're hitting the string codepath for this provided inset though. Looking into this.

nmoucht replied:
Fun `new CSSMathNegate(CSS.px(144))` comes in as a `calc(-144px)` string. Seems like it should be ending up as  a `CSSNumericValue` but maybe this is expected behavior. For now I early return if we get a calc value since I think we need conversion data to be able to convert it to a length.

graouts replied:
Cc @weinig who's been doing work in that area recently.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -74,13 +78,56 @@ Ref<ViewTimeline> ViewTimeline::createFromCSSValue(Style::BuilderState& builderS
-// FIXME: compute offset values from options.
+static std::optional<Length> lengthForInset(std::variant<RefPtr<CSSNumericValue>, RefPtr<CSSKeywordValue>> inset)
+{
+    // TODO: Figure out how to convert CSSNumericValue to Length
```
Isn't that what `lengthForInsetCSSValue()` is for?

nmoucht replied:
No that one does CSSPrimitiveValue -> Length. For this I need to understand how to go from CSSNumericValue -> Length (and how to trigger this codepath via javascript).

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -74,13 +78,56 @@ Ref<ViewTimeline> ViewTimeline::createFromCSSValue(Style::BuilderState& builderS
+    // TODO: Figure out how to convert CSSNumericValue to Length
+    if (holds_alternative<RefPtr<CSSNumericValue>>(inset))
+        auto value = get<RefPtr<CSSNumericValue>>(inset);
+    else if (holds_alternative<RefPtr<CSSKeywordValue>>(inset))
```
This should be just an `else` with an `ASSERT(std::holds_alternative<RefPtr<CSSKeywordValue>>(inset))` since there are only two cases to handle.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -74,13 +78,56 @@ Ref<ViewTimeline> ViewTimeline::createFromCSSValue(Style::BuilderState& builderS
+static std::optional<Length> lengthForInset(std::variant<RefPtr<CSSNumericValue>, RefPtr<CSSKeywordValue>> inset)
+{
+    // TODO: Figure out how to convert CSSNumericValue to Length
+    if (holds_alternative<RefPtr<CSSNumericValue>>(inset))
```
The vast majority of `holds_alternative()` calls use `std::`, I think we should keep doing this here.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -74,13 +78,56 @@ Ref<ViewTimeline> ViewTimeline::createFromCSSValue(Style::BuilderState& builderS
+            return { lengthForInsetCSSValue(dynamicDowncast<CSSPrimitiveValue>(insetPair->protectedFirst())), lengthForInsetCSSValue(dynamicDowncast<CSSPrimitiveValue>(insetPair->protectedSecond())) };
+        return { lengthForInsetCSSValue(dynamicDowncast<CSSPrimitiveValue>(consumedInset)), std::nullopt };
+    }
+    if (holds_alternative<Vector<std::variant<RefPtr<CSSNumericValue>, RefPtr<CSSKeywordValue>>>>(inset)) {
```
This should be an `ASSERT()` since we already returned in the only other scenario.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -74,13 +78,56 @@ Ref<ViewTimeline> ViewTimeline::createFromCSSValue(Style::BuilderState& builderS
+            return { };
+        return { lengthForInset(insetList[0]), insetList.size() == 2 ? std::optional<Length> { lengthForInset(insetList[1]) } : std::nullopt };
+    }
+    return { };
```
Add an ASSERT_NOT_REACHED() here since this code should never be reached.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -74,13 +78,56 @@ Ref<ViewTimeline> ViewTimeline::createFromCSSValue(Style::BuilderState& builderS
+
+static ViewTimelineInsets insetsFromOptions(std::variant<String, Vector<std::variant<RefPtr<CSSNumericValue>, RefPtr<CSSKeywordValue>>>> inset, CSSParserContext context)
+{
+    if (holds_alternative<String>(inset)) {
```
You might consider using `std::get_if` here so that you can both check the type and obtain a pointer if it is the expected type. Saves the `std::get` call below.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -74,13 +78,56 @@ Ref<ViewTimeline> ViewTimeline::createFromCSSValue(Style::BuilderState& builderS
+        auto insetList = get<Vector<std::variant<RefPtr<CSSNumericValue>, RefPtr<CSSKeywordValue>>>>(inset);
+        if (!insetList.size())
+            return { };
+        return { lengthForInset(insetList[0]), insetList.size() == 2 ? std::optional<Length> { lengthForInset(insetList[1]) } : std::nullopt };
```
I think it would read better to split this across to different `return` statements.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.h:
```diff
@@ -55,6 +55,9 @@ class ViewTimeline final : public ScrollTimeline {
     const CSSNumericValue& endOffset() const;
+
+    float startInsetValue() const { return m_insets.start ? m_insets.start->value() : 0; }
+    float endInsetValue() const { return m_insets.end ? m_insets.end->value() : startInsetValue(); }
```
These should be `private` since the only call is within `ViewTimeline`. Even better, you could remove those methods since there is a single call site. Finally, you can use `value_or()` here which will remove the need for the ternary.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -74,13 +78,68 @@ Ref<ViewTimeline> ViewTimeline::createFromCSSValue(Style::BuilderState& builderS
+            auto length = insetValue->convertTo(CSSUnitType::CSS_PX);
+            if (!length)
+                return { };
+            return Length(length->value(), LengthType::Fixed);
```
```suggestion
            if (auto length = insetValue->convertTo(CSSUnitType::CSS_PX))
                return Length(length->value(), LengthType::Fixed);
            return { };
```

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -74,13 +78,68 @@ Ref<ViewTimeline> ViewTimeline::createFromCSSValue(Style::BuilderState& builderS
+        return { lengthForInsetCSSValue(dynamicDowncast<CSSPrimitiveValue>(consumedInset)), std::nullopt };
+    }
+    auto* insetList = std::get_if<Vector<std::variant<RefPtr<CSSNumericValue>, RefPtr<CSSKeywordValue>>>>(&inset);
+    ASSERT(insetList);
```
I'd just `ASSERT(std::holds_alternative())` and use `std::get()`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -74,13 +78,68 @@ Ref<ViewTimeline> ViewTimeline::createFromCSSValue(Style::BuilderState& builderS
+    if (!value || value->isCalculated())
+        return { };
+    if (value->valueID() == CSSValueAuto)
+        return Length();
```
Would be better to stick to `Length()` or `{ }` for consistency.

nmoucht replied:
Fixed.

nmoucht committed:
[scroll-animations] Implement view-timeline-inset property
https://bugs.webkit.org/show_bug.cgi?id=280655
rdar://137009069

Reviewed by Antoine Quint.

Implement parsing of view timeline inset if set via the ViewTimelineOptions,
and adjust computeViewTimelineData to use insets.

* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::lengthForInset):
(WebCore::insetsFromString):
(WebCore::insetsFromOptions):
(WebCore::ViewTimeline::ViewTimeline):
(WebCore::ViewTimeline::computeViewTimelineData const):
(WebCore::ViewTimeline::currentTime):
* Source/WebCore/animation/ViewTimeline.h:

Canonical link: https://commits.webkit.org/284636@main
